java/ql/lib/semmle/code/java/security/UnsafeAndroidAccess.qll

```diff
@@ -26,11 +25,9 @@ abstract class UrlResourceSink extends DataFlow::Node {
  */
 private class CrossOriginUrlResourceSink extends JavaScriptEnabledUrlResourceSink {
   CrossOriginUrlResourceSink() {
-    exists(Variable settings, MethodAccess ma |
-      webViewLoadUrl(this.asExpr(), settings) and
-      ma.getMethod() instanceof CrossOriginAccessMethod and
-      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true and
-      ma.getQualifier() = settings.getAnAccess()
+    exists(WebViewRef webview |
+      webViewLoadUrl(this.asExpr(), webview.getAnAccess()) and
+      isAllowFileAccessEnabled(webview.getAnAccess())
```

Comment:
The double `webview.getAnAccess()` pattern is going to join-order poorly. Push it into each of the two predicates and join on `WebViewRef` instead.

Reply:
Done in [8601137](https://github.com/github/codeql/pull/8537/commits/8601137602d9b29b336b4c89bc2a571567231950).

As I understand it (now that you pointed it out), the number of tuples resulting from both predicates will be way higher because there are (many) more `WebViewRef` accesses than `WebViewRef`s, and in the end the intention is to join on the same `webview` object anyway.

Is this right, or am I missing something?

Reply:
Pretty much. In the double `getAnAccess` code pattern, you are in the worst case asking for all pairs of accesses for each `WebViewRef`. Calculating that set of pairs is going to be quadratic in the average number of `WebViewRef` accesses per `WebViewRef`. Putting the `getAnAccess` inside each of `webViewLoadUrl` and `isAllowFileAccessEnabled` is on the other hand basically free since the step from an access to the corresponding `WebViewRef` is uniquely defined.